File: app-android/app/src/main/java/com/aengix/hopper/model/AppState.kt
package com.aengix.hopper.model

import kotlinx.serialization.Serializable

@Serializable
data class AppState(
    val servers: List<HopNodeProfile> = emptyList(),
    val chains: List<HopChain> = emptyList(),
    val selectedChainID: String? = null,
) {
    val selectedChain: HopChain?
        get() = selectedChainID?.let { id -> chains.firstOrNull { it.id == id } }

    val activeHops: List<HopNodeProfile>
        get() = selectedChain?.let { resolveHops(it) } ?: emptyList()

    val entryHop: HopNodeProfile? get() = activeHops.firstOrNull()

    fun server(id: String): HopNodeProfile? = servers.firstOrNull { it.id == id }

    fun resolveHops(chain: HopChain): List<HopNodeProfile> =
        chain.hopIDs.mapNotNull { server(it) }

    fun addServer(server: HopNodeProfile): AppState = copy(servers = servers + server)

    fun renameServer(id: String, name: String): AppState =
        copy(servers = servers.map { if (it.id == id) it.copy(name = name) else it })

    fun removeServers(ids: Set<String>): AppState {
        val nextServers = servers.filter { it.id !in ids }
        val nextChains = chains.map { chain ->
            chain.copy(hopIDs = chain.hopIDs.filter { it !in ids })
        }
        val nextSelected = selectedChainID?.takeIf { id ->
            nextChains.any { it.id == id }
        } ?: nextChains.firstOrNull()?.id
        return copy(servers = nextServers, chains = nextChains, selectedChainID = nextSelected)
    }

    fun addChain(name: String = ""): Pair<AppState, String> {
        val chain = HopChain(name = name)
        val next = copy(
            chains = chains + chain,
            selectedChainID = selectedChainID ?: chain.id,
        )
        return next to chain.id
    }

    fun removeChains(ids: Set<String>): AppState {
        val nextChains = chains.filter { it.id !in ids }
        val nextSelected = when {
            selectedChainID in ids -> nextChains.firstOrNull()?.id
            else -> selectedChainID
        }
        return copy(chains = nextChains, selectedChainID = nextSelected)
    }

    fun selectChain(id: String?): AppState = copy(selectedChainID = id)

    fun renameChain(id: String, name: String): AppState =
        copy(chains = chains.map { if (it.id == id) it.copy(name = name) else it })

    fun moveHopInChain(chainID: String, fromIndex: Int, toIndex: Int): AppState {
        val chainIndex = chains.indexOfFirst { it.id == chainID }
        if (chainIndex < 0) return this
        val hopIDs = chains[chainIndex].hopIDs.toMutableList()
        if (fromIndex !in hopIDs.indices || toIndex !in 0..hopIDs.size) return this
        val item = hopIDs.removeAt(fromIndex)
        hopIDs.add(toIndex, item)
        val nextChains = chains.toMutableList()
        nextChains[chainIndex] = chains[chainIndex].copy(hopIDs = hopIDs)
        return copy(chains = nextChains)
    }

    fun addServerToChain(chainID: String, serverID: String): AppState {
        if (servers.none { it.id == serverID }) return this
        return copy(chains = chains.map { chain ->
            if (chain.id != chainID || serverID in chain.hopIDs) chain
            else chain.copy(hopIDs = chain.hopIDs + serverID)
        })
    }

    fun removeHopFromChain(chainID: String, indices: Set<Int>): AppState =
        copy(chains = chains.map { chain ->
            if (chain.id != chainID) chain
            else chain.copy(hopIDs = chain.hopIDs.filterIndexed { index, _ -> index !in indices })
        })

    companion object {
        fun fromLegacyHops(hops: List<HopNodeProfile>): AppState {
            var state = AppState(servers = hops)
            if (hops.isEmpty()) return state
            val (next, chainId) = state.addChain(name = "Default")
            state = next.copy(
                chains = next.chains.map { chain ->
                    if (chain.id == chainId) chain.copy(hopIDs = hops.map { it.id }) else chain
                },
                selectedChainID = chainId,
            )
            return state
        }
    }
}

File: app/Shared/AppState.swift
import Foundation

struct AppState: Codable, Equatable {
    var servers: [HopNodeProfile] = []
    var chains: [HopChain] = []
    var selectedChainID: UUID?
    var deployKeys: [DeploySSHKey] = []

    var selectedChain: HopChain? {
        guard let selectedChainID else { return nil }
        return chains.first { $0.id == selectedChainID }
    }

    /// Resolved hops for the selected chain (entry → exit).
    var activeHops: [HopNodeProfile] {
        guard let chain = selectedChain else { return [] }
        return resolveHops(chain)
    }

    var entryHop: HopNodeProfile? { activeHops.first }

    func server(id: UUID) -> HopNodeProfile? {
        servers.first { $0.id == id }
    }

    func resolveHops(_ chain: HopChain) -> [HopNodeProfile] {
        chain.hopIDs.compactMap { server(id: $0) }
    }

    mutating func addServer(_ server: HopNodeProfile) {
        servers.append(server)
    }

    mutating func addDeployKey(_ key: DeploySSHKey) {
        deployKeys.append(key)
    }

    mutating func removeDeployKeys(at offsets: IndexSet) {
        for index in offsets.sorted(by: >) where deployKeys.indices.contains(index) {
            deployKeys.remove(at: index)
        }
    }

    func deployKey(id: UUID) -> DeploySSHKey? {
        deployKeys.first { $0.id == id }
    }

    mutating func renameServer(id: UUID, name: String) {
        guard let index = servers.firstIndex(where: { $0.id == id }) else { return }
        servers[index].name = name
    }

    mutating func removeServers(at offsets: IndexSet) {
        let removed = offsets.compactMap { servers.indices.contains($0) ? servers[$0].id : nil }
        for index in offsets.sorted(by: >) where servers.indices.contains(index) {
            servers.remove(at: index)
        }
        for i in chains.indices {
            chains[i].hopIDs.removeAll { removed.contains($0) }
        }
        if let selected = selectedChainID, !chains.contains(where: { $0.id == selected }) {
            selectedChainID = chains.first?.id
        }
    }

    mutating func addChain(name: String = "") -> UUID {
        let chain = HopChain(name: name)
        chains.append(chain)
        if selectedChainID == nil {
            selectedChainID = chain.id
        }
        return chain.id
    }

    mutating func removeChains(at offsets: IndexSet) {
        let removed = offsets.compactMap { chains.indices.contains($0) ? chains[$0].id : nil }
        for index in offsets.sorted(by: >) where chains.indices.contains(index) {
            chains.remove(at: index)
        }
        if let selected = selectedChainID, removed.contains(selected) {
            selectedChainID = chains.first?.id
        }
    }

    mutating func selectChain(_ id: UUID?) {
        selectedChainID = id
    }

    mutating func renameChain(id: UUID, name: String) {
        guard let index = chains.firstIndex(where: { $0.id == id }) else { return }
        chains[index].name = name
    }

    mutating func moveHopInChain(chainID: UUID, from source: IndexSet, to destination: Int) {
        guard let index = chains.firstIndex(where: { $0.id == chainID }) else { return }
        var hopIDs = chains[index].hopIDs
        let moving = source.sorted().map { hopIDs[$0] }
        var target = destination
        for offset in source.sorted() where offset < destination {
            target -= 1
        }
        for offset in source.sorted(by: >) {
            hopIDs.remove(at: offset)
        }
        for (i, id) in moving.enumerated() {
            hopIDs.insert(id, at: target + i)
        }
        chains[index].hopIDs = hopIDs
    }

    mutating func addServerToChain(chainID: UUID, serverID: UUID) {
        guard let index = chains.firstIndex(where: { $0.id == chainID }) else { return }
        guard servers.contains(where: { $0.id == serverID }) else { return }
        guard !chains[index].hopIDs.contains(serverID) else { return }
        chains[index].hopIDs.append(serverID)
    }

    mutating func removeHopFromChain(chainID: UUID, at offsets: IndexSet) {
        guard let index = chains.firstIndex(where: { $0.id == chainID }) else { return }
        for offset in offsets.sorted(by: >) where chains[index].hopIDs.indices.contains(offset) {
            chains[index].hopIDs.remove(at: offset)
        }
    }

    /// Migrate flat hop list from older app versions.
    static func fromLegacyHops(_ hops: [HopNodeProfile]) -> AppState {
        var state = AppState(servers: hops)
        if hops.isEmpty { return state }
        let chainID = state.addChain(name: "Default")
        if let chainIndex = state.chains.firstIndex(where: { $0.id == chainID }) {
            state.chains[chainIndex].hopIDs = hops.map(\.id)
        }
        state.selectedChainID = chainID
        return state
    }
}
